libcoap/src/types.rs

```diff
@@ -258,153 +270,545 @@ impl From<coap_uri_scheme_t> for CoapUriScheme {
     }
 }
 
-/// Representation of the host part of a CoAP request.
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub enum CoapUriHost {
-    IpLiteral(IpAddr),
-    Name(String),
-}
-
-impl<T: ToString> From<url::Host<T>> for CoapUriHost {
-    fn from(host: Host<T>) -> Self {
-        match host {
-            Host::Domain(d) => CoapUriHost::Name(d.to_string()),
-            Host::Ipv4(addr) => CoapUriHost::IpLiteral(IpAddr::V4(addr)),
-            Host::Ipv6(addr) => CoapUriHost::IpLiteral(IpAddr::V6(addr)),
+impl From<CoapProtocol> for CoapUriScheme {
+    fn from(value: CoapProtocol) -> Self {
+        match value {
+            CoapProtocol::None | CoapProtocol::Udp => CoapUriScheme::Coap,
+            CoapProtocol::Dtls => CoapUriScheme::Coaps,
+            CoapProtocol::Tcp => CoapUriScheme::CoapTcp,
+            CoapProtocol::Tls => CoapUriScheme::CoapsTcp,
         }
     }
 }
 
-impl FromStr for CoapUriHost {
-    type Err = Infallible;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Ok(IpAddr::from_str(s).map_or_else(|_| CoapUriHost::Name(s.to_string()), CoapUriHost::IpLiteral))
-    }
+/// Representation of a URI for CoAP requests, responses or proxy URIs.
+///
+/// See https://datatracker.ietf.org/doc/html/rfc7252#section-6 for a description of how a URI
+/// should look like.
+///
+/// # Examples
+/// The easiest way to instantiate a request or location CoAP URI is by parsing a string (either
+/// using the [FromStr] implementation or using [CoapUri::try_from_str]):
+/// ```
+/// use libcoap_rs::error::UriParsingError;
+/// use libcoap_rs::types::{CoapUri, CoapUriScheme};
+///
+/// let uri: CoapUri = "coap://example.com:4711/foo/bar?answer=42".parse()?;
+///
+/// assert_eq!(uri.scheme(), Some(CoapUriScheme::Coap));
+/// assert_eq!(uri.host(), Some("example.com".as_bytes()));
+/// assert_eq!(uri.port(), Some(4711));
+/// assert_eq!(uri.path(), Some("foo/bar".as_bytes()));
+/// assert_eq!(uri.query(), Some("answer=42".as_bytes()));
+/// assert!(!uri.is_proxy());
+///
+/// # Result::<(), UriParsingError>::Ok(())
+/// ```
+///
+/// Alternatively, a [CoapUri] may be constructed from its parts using [CoapUri::new] or
+/// [CoapUri::new_relative] or from a [Url] (requires the `url` feature), refer to the method level
+/// documentation for more information.
+///
+/// If you want to create a proxy URI, refer to the method-level documentation [CoapUri::new_proxy],
+/// [CoapUri::try_from_str_proxy] or [CoapUri::try_from_url_proxy].
+///
+/// # Note on URI Length Limits
+///
+/// Due to [the specified limits](https://datatracker.ietf.org/doc/html/rfc7252#section-5.10)
+/// of CoAP option lengths, URI path components for a request/location URI must not exceed 255
+/// characters each, i.e. a full path with more than 255 characters is fine, but each individual
+/// path segment must be smaller than 255 bytes.
```

Comment:
There is a difference between "must not exceed 255 characters" and "must be smaller than 255 bytes". This is still consistent if the latter includes the null byte, but I think being more explicit here wouldn't hurt.

Reply:
I think it would be best if all instances in this paragraph referred to either "characters" or "bytes", but not both. I know what's meant by this now, but I think an outside reader could still get confused by this in some cases.